File: examples/go/onboard.go
// Package main implements a basic OpenConstruct agent onboarding example in Go.
//
// # OpenConstruct — Agent Onboarding (Go)
//
// Registers an agent with the OpenConstruct coordinator and maintains
// a heartbeat loop using the official Go SDK.
//
// ## Usage
//
//	go run onboard.go --coordinator ws://localhost:9142
package main

import (
	"context"
	"flag"
	"log/slog"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/openconstruct/go-sdk"
)

func main() {
	// ── Flags ────────────────────────────────────────────────────────
	coordinator := flag.String("coordinator", "ws://localhost:9142", "Coordinator WebSocket URL")
	name := flag.String("name", "go-agent", "Agent name")
	flag.Parse()

	// ── Structured logger ────────────────────────────────────────────
	logger := slog.New(slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{Level: slog.LevelInfo}))
	slog.SetDefault(logger)

	// ── Connect ──────────────────────────────────────────────────────
	client, err := openconstruct.Connect(*coordinator)
	if err != nil {
		slog.Error("failed to connect to coordinator", "error", err)
		os.Exit(1)
	}
	slog.Info("connected to coordinator", "url", *coordinator)

	// ── Declare identity and capabilities ────────────────────────────
	identity := openconstruct.AgentIdentity{
		Name:         *name,
		Capabilities: []openconstruct.Capability{openconstruct.CapSense, openconstruct.CapAct},
		Metadata: map[string]string{
			"language": "go",
			"example":  "onboard",
		},
	}

	// ── Onboard ──────────────────────────────────────────────────────
	session, err := client.Onboard(context.Background(), identity)
	if err != nil {
		slog.Error("onboarding failed", "error", err)
		os.Exit(1)
	}
	slog.Info("onboarded",
		"agent_id", session.AgentID(),
		"room", session.AssignedRoom(),
	)

	// ── Graceful shutdown ────────────────────────────────────────────
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		sig := <-sigCh
		slog.Info("received signal, shutting down", "signal", sig)
		cancel()
	}()

	// ── Heartbeat loop ───────────────────────────────────────────────
	ticker := time.NewTicker(30 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			session.Disconnect()
			slog.Info("disconnected — bye")
			return
		case <-ticker.C:
			status, err := session.Heartbeat(ctx)
			if err != nil {
				slog.Warn("heartbeat failed", "error", err)
				continue
			}
			slog.Info("heartbeat acknowledged", "status", status)
		}
	}
}

File: examples/rust/onboard.rs
//! # OpenConstruct — Basic Agent Onboarding (Rust)
//!
//! Demonstrates how an agent registers itself with the OpenConstruct
//! coordinator, receives its assigned room topology, and enters the
//! heartbeat loop.
//!
//! ## Usage
//! ```sh
//! cargo run --example onboard -- --coordinator ws://localhost:9142
//! ```

use anyhow::{Context, Result};
use clap::Parser;
use openconstruct::{
    AgentConfig, AgentIdentity, Capability, CoordinatorClient, HeartbeatConfig,
};
use std::time::Duration;
use tracing::{info, warn};

/// Command-line arguments for the onboarding example.
#[derive(Parser)]
#[command(name = "onboard", about = "Basic agent onboarding example")]
struct Args {
    /// WebSocket URL of the OpenConstruct coordinator.
    #[arg(long, default_value = "ws://localhost:9142")]
    coordinator: String,

    /// Human-readable name for this agent.
    #[arg(long, default_value = "example-agent")]
    name: String,
}

#[tokio::main]
async fn main() -> Result<()> {
    // ── Initialise structured logging ───────────────────────────────
    tracing_subscriber::fmt()
        .with_env_filter("openconstruct=debug,info")
        .init();

    let args = Args::parse();

    // ── Build the coordinator client ────────────────────────────────
    let client = CoordinatorClient::connect(&args.coordinator)
        .await
        .context("failed to connect to coordinator")?;
    info!(url = %args.coordinator, "connected to coordinator");

    // ── Declare our identity and capabilities ───────────────────────
    let identity = AgentIdentity {
        name: args.name.clone(),
        capabilities: vec![Capability::Sense, Capability::Act],
        metadata: serde_json::json!({
            "language": "rust",
            "example": "onboard",
        }),
    };

    // ── Onboard ─────────────────────────────────────────────────────
    let config = AgentConfig {
        heartbeat: HeartbeatConfig {
            interval: Duration::from_secs(30),
            timeout: Duration::from_secs(10),
        },
        ..Default::default()
    };

    let session = client
        .onboard(identity, config)
        .await
        .context("onboarding handshake failed")?;

    info!(
        agent_id = %session.agent_id(),
        room = ?session.assigned_room(),
        "onboarding complete — assigned to room"
    );

    // ── Heartbeat loop (runs until interrupted) ─────────────────────
    loop {
        match session.heartbeat().await {
            Ok(status) => {
                info!(status = ?status, "heartbeat acknowledged");
            }
            Err(e) => {
                warn!(error = %e, "heartbeat failed — will retry");
            }
        }
        tokio::time::sleep(Duration::from_secs(30)).await;
    }
}
